### PR TITLE
synchronise resource tree modification

### PR DESCRIPTION
**Pull request by gireeshpunathil**
SearchablePluginsManager saves the state of the resources in the workspace without acquiring a lock, or not scheduling the job asynchronously.

Fix it by employing the WorkspaceJob scheduler.

Fixes: #64

### CODE REVIEW

**Comment by gireeshpunathil** [comment]
I have made a number of assumptions in this PR, which I would like to be validated as part of the review. thanks in advance to whover is going to do that!

- workspace-wide resource modification is delegated to `org.eclipse.core.resources.WorkspaceJob`. Is this the right job for this type of work?
- the rule of this job is set to be the root of the resource, following how it is done elsewhere in the code. Is there a place where it defines what type of rules apply to what type of jobs?
- to bring up the progressive status, I am using Display class, which introduces a new bundle-lvele dependency with `oreg.eclipse.ui`. Is it normal, safe, and non-breaking? do I need to worry about this anywhere - such as bumping the plugin version or something?

**Comment by vik-chand** [comment]
@iloveeclipse can you please review this pr?

**Comment by merks** [comment]
The core should not depend on the UI! 

I think you could use one of these instead of scheduling a job:

org.eclipse.core.resources.IWorkspace.run(ICoreRunnable, ISchedulingRule, int, IProgressMonitor)
org.eclipse.core.resources.IWorkspace.run(IWorkspaceRunnable, ISchedulingRule, int, IProgressMonitor)

org.eclipse.pde.internal.core.BundleValidationOperation uses that approach.

**Comment by iloveeclipse** [comment]
> @iloveeclipse can you please review this pr?

Will try, have few other PR's in the review list.

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SearchablePluginsManager.java [hunk 1]
```diff
@@ -400,37 +406,52 @@ public void removePluginModelListener(IPluginModelListener listener) {
...
 			}
-		}
+				return Status.OK_STATUS;
+			}
+		};
+			job.setSystem(true);
+			job.setRule(ResourcesPlugin.getWorkspace().getRoot());
+			Display display = Display.getCurrent();
+			if (display != null) {
+				PlatformUI.getWorkbench().getProgressService().showInDialog(display.getActiveShell(),
```
Comment (iloveeclipse) [inline ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SearchablePluginsManager.java:451]: Please don't use that API (anymore). We won't too many explicit popups, especially here for that operation

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/pderesources.properties [hunk 2]
```diff
@@ -256,6 +256,7 @@ XMLTextChangeListener_editNames_modifyAttribute=Modify "{0}" attribute of "{1}"
 XMLTextChangeListener_editNames_addAttribute=Add "{0}" attribute to "{1}" node
 
 SearchablePluginsManager_createProjectTaskName=Creating {0} project
+SearchablePluginsManager_SaveStates=save states
```
Comment (iloveeclipse) [inline ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/pderesources.properties:259]: Please check messages used and note the new one uses inconsistent style and wording.

Comment (gireeshpunathil) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/pderesources.properties]: same as above, removed.

**Comment by gireeshpunathil** [comment]
@merks @iloveeclipse - pushed new change with the approach suggested, pls have a look; thanks!

**Review by iloveeclipse — CHANGES_REQUESTED** [review]
General note: looks now better in code, haven't validated if the problem disappears.

- Please **rebase** on latest master state (do not merge!)
- Please bump last bundle version segment with +100.

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SearchablePluginsManager.java [hunk 3]
```diff
@@ -400,36 +404,51 @@ public void removePluginModelListener(IPluginModelListener listener) {
...
 					}
+				} catch (IOException e) {
+					PDECore.log(e);
 				}
-			} catch (IOException | CoreException e) {
-				PDECore.log(e);
 			}
+		};
+		try {
+			if (workspace.isTreeLocked()) {
```
[the remark this answers is not in the document]
Comment (gireeshpunathil) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SearchablePluginsManager.java:442]: @iloveeclipse - as per my understanding, the `workspace` lock seems re-enterable, and if it is already locked (by the same thread), you could just go ahead and run with no rules and restrictions, which is light-weight, than its heavy-weight  counterpart. Isn't that the case?

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SearchablePluginsManager.java [hunk 4]
```diff
@@ -400,36 +404,51 @@ public void removePluginModelListener(IPluginModelListener listener) {
...
+				runnable.run(null);
+			} else {
+				IResourceRuleFactory factory = workspace.getRuleFactory();
+				IWorkspaceRoot root = workspace.getRoot();
+				ISchedulingRule rule = factory.modifyRule(root);
+				workspace.run(runnable, rule, IWorkspace.AVOID_UPDATE, null);
```
[the remark this answers is not in the document]
Comment (gireeshpunathil) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SearchablePluginsManager.java:448]: I have been trying to figure out a way to obtain a monitor in the current context, but could not - do you know? I have been largely following this block of code which does a similar thing:

https://github.com/eclipse-platform/eclipse.platform/blob/9839c18f52687db6f949bdb56a59a5d20cd7c938/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ProjectPreferences.java#L659-L734

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SearchablePluginsManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (iloveeclipse) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SearchablePluginsManager.java:442]: OK, so it is just optimization, not something you **needed** here - that was my question.

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SearchablePluginsManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (iloveeclipse) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SearchablePluginsManager.java:448]: Arrgh, sorry, I've assumed the monitor is there looking at line 408 in the diff. 
No, there is no outer monitor here, so it should be fine fine as is.

**Review by iloveeclipse — COMMENTED** [review]
I haven't validated, but my questions were answered now, so should in theory work.

**Comment by gireeshpunathil** [comment]
there are not outstanding review comments, so this should be ready to merge, has been in this state for a while.

**Comment by iloveeclipse** [comment]
> there are not outstanding review comments, so this should be ready to merge, has been in this state for a while.

Sorry, too many bugs & reviews...

**Comment by jukzi** [comment]
@gireeshpunathil  thanks for fixing, @iloveeclipse thanks for review

**Pull request merged**